tests/plugins/Qukeys/basic/test/testcase.cpp

```diff
@@ -32,182 +32,248 @@ constexpr KeyAddr key_addr_X{3, 2};
+                                   std::string message = "") {
+    sim_.RunCycles(n);
+    return expectReport(added_keys, removed_keys, message);
+  }
+  uint32_t expectReportAfterCycles(size_t n, std::string message = "") {
+    return expectReportAfterCycles(n, std::set<Key>{}, std::set<Key>{},
+                                   message);
+  }
+  uint32_t expectReportAfterCyclesWith(size_t n, Key key,
+                                       std::string message = "") {
+    sim_.RunCycles(n);
+    return expectReportWith(key, message);
+  }
+  uint32_t expectReportAfterCyclesWithout(size_t n, Key key,
+                                          std::string message = "") {
+    sim_.RunCycles(n);
+    return expectReportWithout(key, message);
+  }
+  uint32_t expectReportAfterMillis(uint32_t t,
+                                   std::set<Key> added_keys,
+                                   std::set<Key> removed_keys,
+                                   std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReport(added_keys, removed_keys, message);
+  }
+  uint32_t expectReportAfterMillis(uint32_t t, std::string message = "") {
+    return expectReportAfterMillis(t, std::set<Key>{}, std::set<Key>{},
+                                   message);
+  }
+  uint32_t expectReportAfterMillisWith(uint32_t t, Key key,
+                                       std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReportWith(key, message);
+  }
+  uint32_t expectReportAfterMillisWithout(uint32_t t, Key key,
+                                          std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReportWithout(key, message);
+  }
 };
 
 TEST_F(QukeysBasic, TapQukeyAlone) {
 
-  // Press `A`
-  sim_.Press(key_addr_A);
-
-  state_ = RunCycle();
-
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 0)
-      << "There should be no HID report after the qukey is pressed";
-  sim_.RunForMillis(20);
-  // Release `A`
-  sim_.Release(key_addr_A);
-
-  sim_.RunCycles(2);
-  state_ = RunCycle();
-
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 2)
-      << "There should be two HID reports after the release of a tapped qukey";
-
-  expected_keycodes_.insert(Key_A.getKeyCode());
-  EXPECT_THAT(state_->HIDReports()->Keyboard(0).ActiveKeycodes(),
-              ::testing::ElementsAreArray(expected_keycodes_))
-      << "The first report should include only `A`";
-
-  expected_keycodes_.erase(Key_A.getKeyCode());
-  EXPECT_THAT(state_->HIDReports()->Keyboard(1).ActiveKeycodes(),
-              ::testing::ElementsAreArray(expected_keycodes_))
-      << "The second report should be empty";
-
-  state_ = RunCycle();
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 0);
+  pressKey(10, key_addr_A);
+  releaseKey(50, key_addr_A);
+  expectReportAfterCyclesWith(
+      1, Key_A, "Report should contain only `A`");
+  expectReportAfterCyclesWithout(
+      1, Key_A, "Report should be empty");
```

Comment:
This is now pretty close to what I'm aiming for testcases to look like (at least, the ones primarily concerned with keyswitch input & USB output).  This testcase is very simple; it simulates a single key press, followed by a release, followed by two keyboard HID reports, in that order. In each case, the first parameter is the number of milliseconds or cycles to run the simulator before the event. The `pressKey()` & `releaseKey()` methods specify the keyswitch affected. The `expectReport*()` methods specify changes to the report (in terms of `Key` values for readability), and a message to display on failure. All of the test verification is done afterwards, to make it a simple as possible to follow what the test is doing.